extras/module-node/README.rst

```diff
@@ -0,0 +1,54 @@
+=============================
+Node.js-like module framework
+=============================
+
+This directory contains an example module resolution and loading framework and
+``require`` implementation based on Node.js.  The application need only
+provide the module resolution and loading logic:
+
+* Add ``duk_module_node.c`` to list of C sources to compile.
+
+* Ensure ``duk_module_node.h`` is in the include path.
+
+* Include the extra header in calling code and initialize the bindings::
+
+      #include "duktape.h"
+      #include "duk_module_node.h"
+
+      /* After initializing the Duktape heap or when creating a new
+       * thread with a new global environment:
+       */
+      duk_module_node_init(ctx, resolve_cb, load_cb);
+
+* The resolve callback looks like this::
+
+      duk_ret_t cb_resolve_module(duk_context *ctx) {
+          const char *parent_id = duk_get_string(ctx, -2);  /* calling module ID */
+          const char *module_id = duk_get_string(ctx, -1);
+
+          /* Arrive at the canonical module ID in whatever way makes sense */
+
+          duk_push_string(ctx, resolved_id);
+          return 1; /*nrets*/
+      }
+
+  Important note: When ``require()`` is called outside of any module, the
+  parent ID will be an empty string.
+
+* The load callback looks like this::
+
+      duk_ret_t cb_load_module(duk_context *ctx) {
+          /* `module` object on stack top, `module.id` is the resolved ID */
+
+          /* Retrieve the module's JS source code */
+
+          duk_push_string(ctx, source_string);
+          return 1; /*nrets*/
```

Comment:
Trivia: typically there are two spaces between a semicolon and a same-line comment (just for some visual separation).